File: solutions/typescript/dnd-character/1/dnd-character.ts
export class DnDCharacter {
    strength: number;
    dexterity: number;
    constitution: number;
    intelligence: number;
    wisdom: number;
    charisma: number;
    hitpoints: number;

    constructor() {
        this.strength = DnDCharacter.generateAbilityScore();
        this.dexterity = DnDCharacter.generateAbilityScore();
        this.constitution = DnDCharacter.generateAbilityScore();
        this.intelligence = DnDCharacter.generateAbilityScore();
        this.wisdom = DnDCharacter.generateAbilityScore();
        this.charisma = DnDCharacter.generateAbilityScore();
        this.hitpoints = 10 + DnDCharacter.getModifierFor(this.constitution);
    }

    private static roll(sides: number): number {
        return Math.ceil(Math.random() * sides);
    }

    public static generateAbilityScore(): number {
        const rolls = Array.from({ length: 4 }, () => this.roll(6));

        const min = Math.min(...rolls);

        for (const i of rolls) {
            const roll = rolls[i];

            if (min === roll) {
                rolls.splice(i, 1);
                break;
            }
        }

        return rolls.reduce((a, b) => a + b, 0);
    }

    public static getModifierFor(abilityValue: number): number {
        return Math.floor((abilityValue - 10) / 2);
    }
}

File: solutions/csharp/dnd-character/1/DndCharacter.cs
public class DndCharacter
{
    public int Strength { get; init;  }
    public int Dexterity { get; init;  }
    public int Constitution { get; init;  }
    public int Intelligence { get; init;  }
    public int Wisdom { get; init;  }
    public int Charisma { get; init;  }
    public int Hitpoints { get; init;  }

    public static int Modifier(int score) => (int)Math.Floor(((double)score - 10) / 2);

    public static int Ability() => Roll.D(4, 6)
        .OrderDescending()
        .Take(3)
        .Sum();

    public static DndCharacter Generate() {
        var constitution = Ability();
        
        return new DndCharacter
        {
            Strength = Ability(),
            Dexterity = Ability(),
            Constitution = constitution,
            Intelligence = Ability(),
            Wisdom = Ability(),
            Charisma = Ability(),
            Hitpoints = 10 + Modifier(constitution),
        };
    }

}

internal static class Roll
{
    private static readonly Random Rng = new();

    private static int D(int die) => Rng.Next(1, die + 1);

    public static int[] D(int n, int die) => Enumerable.Range(0, n)
        .Select(_ => D(die))
        .ToArray();
}
